src/state/slices/opportunitiesSlice/resolvers/foxfarming.ts

```diff
@@ -122,6 +138,104 @@ export const foxFarmingLpMetadataResolver = async ({
 
   return { data }
 }
+
+export const foxFarmingStakingMetadataResolver = async ({
+  opportunityId,
+  opportunityType,
+  reduxApi,
+}: {
+  opportunityId: LpId | StakingId
+  opportunityType: OpportunityDefiType
+  reduxApi: ReduxApi
+}): Promise<{ data: GetOpportunityMetadataOutput }> => {
+  const { getState } = reduxApi
+  const state: any = getState() // ReduxState causes circular dependency
+  const assets: AssetsState = state.assets
+  const lpAssetPrecision = assets.byId[foxEthLpAssetId].precision
+  const foxPrecision = assets.byId[foxAssetId].precision
+  const ethPrecision = assets.byId[ethAssetId].precision
+  const lpTokenMarketData: MarketData = selectMarketDataById(state, foxEthLpAssetId)
+  const lpTokenPrice = lpTokenMarketData?.price
+
+  const { assetReference: contractAddress } = fromAssetId(opportunityId as AssetId)
+
+  if (bnOrZero(lpTokenPrice).isZero()) {
+    throw new Error(`Market data not ready for ${foxEthLpAssetId}`)
+  }
+
+  const ethersProvider = getEthersProvider()
+  const foxFarmingContract = getOrCreateContract(contractAddress, farmingAbi)
+  const uniV2LPContract = getOrCreateContract(UNISWAP_V2_WETH_FOX_POOL_ADDRESS, IUniswapV2Pair.abi)
+
+  // tvl
+  const totalSupply = await foxFarmingContract.totalSupply()
+  const tvl = bnOrZero(totalSupply.toString())
+    .div(bn(10).pow(lpAssetPrecision))
+    .times(lpTokenPrice)
+    .toFixed(2)
+
+  // apr
+  const foxRewardRatePerTokenV4 = await rewardRatePerToken(foxFarmingContract)
+  const pair = await Fetcher.fetchPairData(
+    new Token(0, WETH_TOKEN_CONTRACT_ADDRESS, ethPrecision),
+    new Token(0, FOX_TOKEN_CONTRACT_ADDRESS, foxPrecision),
+    ethersProvider,
+  )
+
+  // Getting the ratio of the LP token for each asset
+  // fetchPairData().reserve0 and reserve1 somehow return crypto human amounts as opposed to getReserves() using full precision notation
+  const foxReserves = toBaseUnit(
+    bnOrZero(bnOrZero(pair.reserve1.toFixed()).toString()).toString(),
+    pair.token1.decimals,
```

Comment:
Baby steps into abstracting this logic to work with any UniV2 opportunity, the guts of this module will work for any opportunity if we replace `ETH` and `FOX` terminologies by `token0` and `token`